File: electron/src/compression.ts
import { gzipSync, gunzipSync } from "zlib";

const COMPRESSION_VERSION = 1;
const COMPRESSION_TYPE = "gzip";

export type CompressedEnvelope = {
  v: number;
  compression: string;
  payload: string;
};

/**
 * Compresses a plaintext string using gzip and wraps it in a versioned envelope.
 *
 * @param plaintext The string to compress
 * @returns JSON string containing compressed payload
 */
export function compressPayload(plaintext: string): string {
  const compressed = gzipSync(Buffer.from(plaintext, "utf-8"));
  const base64 = compressed.toString("base64");
  const envelope: CompressedEnvelope = {
    v: COMPRESSION_VERSION,
    compression: COMPRESSION_TYPE,
    payload: base64,
  };
  return JSON.stringify(envelope);
}

/**
 * Decompresses a payload envelope back to plaintext.
 *
 * Handles both compressed envelopes and raw plaintext (for backwards compatibility).
 *
 * @param envelope The JSON envelope or raw plaintext
 * @returns Decompressed plaintext string
 */
export function decompressPayload(envelope: string): string {
  const trimmed = envelope.trim();
  if (!trimmed.startsWith("{")) {
    return envelope; // raw plaintext
  }

  try {
    const parsed = JSON.parse(trimmed);
    if (!parsed.compression || !parsed.payload) {
      return envelope; // not a compression envelope
    }

    const compressed = Buffer.from(parsed.payload, "base64");
    const decompressed = gunzipSync(compressed);
    return decompressed.toString("utf-8");
  } catch {
    return envelope; // parse error, treat as raw
  }
}

/**
 * Checks if a payload should be compressed based on size.
 *
 * @param plaintext The string to check
 * @param threshold Minimum size in bytes to trigger compression (default 1KB)
 * @returns true if the payload should be compressed
 */
export function shouldCompress(plaintext: string, threshold = 1024): boolean {
  return Buffer.byteLength(plaintext, "utf-8") >= threshold;
}

/**
 * Compresses payload only if it exceeds the size threshold.
 *
 * @param plaintext The string to potentially compress
 * @param threshold Minimum size in bytes to trigger compression (default 1KB)
 * @returns Compressed envelope if above threshold, otherwise raw plaintext
 */
export function maybeCompressPayload(plaintext: string, threshold = 1024): string {
  return shouldCompress(plaintext, threshold) ? compressPayload(plaintext) : plaintext;
}

File: android/bugstr-nostr-crypto/src/main/kotlin/com/bugstr/nostr/crypto/Compression.kt
package com.bugstr.nostr.crypto

import android.util.Base64
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

private const val COMPRESSION_VERSION = 1
private const val COMPRESSION_TYPE = "gzip"

/**
 * Compresses a plaintext string using gzip and wraps it in a versioned envelope.
 *
 * Output format: {"v":1,"compression":"gzip","payload":"<base64>"}
 *
 * @param plaintext The string to compress
 * @return JSON envelope containing compressed payload
 */
fun compressPayload(plaintext: String): String {
    val compressed = ByteArrayOutputStream().use { baos ->
        GZIPOutputStream(baos).use { gzip ->
            gzip.write(plaintext.toByteArray(Charsets.UTF_8))
        }
        baos.toByteArray()
    }
    val base64 = Base64.encodeToString(compressed, Base64.NO_WRAP)
    return buildString {
        append("{\"v\":")
        append(COMPRESSION_VERSION)
        append(",\"compression\":\"")
        append(COMPRESSION_TYPE)
        append("\",\"payload\":\"")
        append(base64)
        append("\"}")
    }
}

/**
 * Decompresses a payload envelope back to plaintext.
 *
 * Handles both compressed envelopes and raw plaintext (for backwards compatibility).
 *
 * @param envelope The JSON envelope or raw plaintext
 * @return Decompressed plaintext string
 */
fun decompressPayload(envelope: String): String {
    val trimmed = envelope.trim()
    if (!trimmed.startsWith("{") || !trimmed.contains("\"compression\"")) {
        return envelope // raw plaintext, no compression
    }

    // Simple JSON parsing without dependencies
    val payloadStart = trimmed.indexOf("\"payload\":\"") + 11
    if (payloadStart < 11) return envelope
    val payloadEnd = trimmed.indexOf("\"", payloadStart)
    if (payloadEnd < 0) return envelope

    val base64 = trimmed.substring(payloadStart, payloadEnd)
    val compressed = Base64.decode(base64, Base64.NO_WRAP)

    return ByteArrayInputStream(compressed).use { bais ->
        GZIPInputStream(bais).use { gzip ->
            gzip.bufferedReader(Charsets.UTF_8).readText()
        }
    }
}

/**
 * Checks if a payload should be compressed based on size.
 *
 * Small payloads may not benefit from compression overhead.
 *
 * @param plaintext The string to check
 * @param threshold Minimum size in bytes to trigger compression (default 1KB)
 * @return true if the payload should be compressed
 */
fun shouldCompress(plaintext: String, threshold: Int = 1024): Boolean {
    return plaintext.toByteArray(Charsets.UTF_8).size >= threshold
}

/**
 * Compresses payload only if it exceeds the size threshold.
 *
 * @param plaintext The string to potentially compress
 * @param threshold Minimum size in bytes to trigger compression (default 1KB)
 * @return Compressed envelope if above threshold, otherwise raw plaintext
 */
fun maybeCompressPayload(plaintext: String, threshold: Int = 1024): String {
    return if (shouldCompress(plaintext, threshold)) {
        compressPayload(plaintext)
    } else {
        plaintext
    }
}
